File: 2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 //每次循环都要new一个新结点,不要妄想直接改变l1或者l2的值作为返回值,因为有进n位的可能甚至相加的位数比l1,l2都长.
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *move1 = l1;
        ListNode *move2 = l2;
        ListNode *root = new ListNode(-1);//先设置root结点指向返回的链表的首结点
        ListNode *move3 = root;//用一个move3做辅助为新产生的结点用next连接起来
        int mark = 0;
        while (move1 || move2 || mark == 1) {
            ListNode *new_node = new ListNode(0);
            move3->next = new_node;//每产生一个node,都把上一个结点指向它
            int move1_val = move1 ? move1->val : 0;//有可能move1,move2是nullptr
            int move2_val = move2 ? move2->val : 0;
            new_node->val = ((move1_val + move2_val) + mark) % 10;
            if (((move1_val + move2_val)  + mark) >= 10) {
                mark = 1;
            } else {
                mark = 0;
            }
            
            move1 = move1 ? move1->next : move1;//这里要注意,因为有可能下一为是nullptr
            move2 = move2 ? move2->next : move2;
            move3 = move3->next;
        }
        return root->next;
    }
};  

File: 2.go
//https://leetcode.com/problems/add-two-numbers
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    head := &ListNode{Val: -1, Next: nil} //头结点无实际意义，只是为了返回第一个节点
    curNode := head;
    carry := 0 //进位
    for l1 != nil || l2 != nil || carry == 1 {
        left := 0
        if l1 != nil {
            left, l1 = l1.Val, l1.Next
        }
        right := 0
        if l2 != nil {
            right, l2 = l2.Val, l2.Next
        }
        
        curVal := left + right + carry
    
        carry = 0
        if (curVal - 10 >= 0) {
            curVal -= 10
            carry = 1
        }
        
        curNode.Next = &ListNode{Val: curVal, Next: nil}
        curNode = curNode.Next

    }
    return head.Next
}
